ariac_plugins/src/assembly_state_publisher.cpp

```diff
@@ -0,0 +1,181 @@
+  bool sensor_attached_;
+
+  gazebo::transport::NodePtr gznode_;
+  gazebo::transport::SubscriberPtr battery_sub_;
+  gazebo::transport::SubscriberPtr pump_sub_;
+  gazebo::transport::SubscriberPtr regulator_sub_;
+  gazebo::transport::SubscriberPtr sensor_sub_;
+
+  gazebo::common::Timer timer_;
+  int rate_;
+
+  void PublishState();
+};
+
+AssemblyStatePublisher::AssemblyStatePublisher()
+: impl_(std::make_unique<AssemblyStatePublisherPrivate>())
+{
+}
+
+AssemblyStatePublisher::~AssemblyStatePublisher()
+{
+}
+
+void AssemblyStatePublisher::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf)
+{
+  impl_->model_ = model;
+  std::string model_name = model->GetName();
+  std::string link_name = "insert_link";
+
+  /* Causes AGV controllers to fail to load? */
+  sdf::ElementPtr ns(new sdf::Element);
+  sdf->SetParent(sdf->FindElement("ros"));
+  ns->SetName("namespace");
+  ns->AddValue("string", model_name, "1");
+  sdf->FindElement("ros")->InsertElement(ns);
+
+  impl_->ros_node_ = gazebo_ros::Node::Get(sdf);
+
+  std::string bat_topic = "/gazebo/world/" + model_name + "/" + link_name + "/battery_attached";
+  std::string pum_topic = "/gazebo/world/" + model_name + "/" + link_name + "/pump_attached";
+  std::string reg_topic = "/gazebo/world/" + model_name + "/" + link_name + "/regulator_attached";
+  std::string sen_topic = "/gazebo/world/" + model_name + "/" + link_name + "/sensor_attached";
+
+  // Initialize a gazebo node and subscribe to each assembly topic
+  impl_->gznode_ = gazebo::transport::NodePtr(new gazebo::transport::Node());
+  impl_->gznode_->Init(impl_->model_->GetWorld()->Name());
+
+  impl_->battery_sub_ = impl_->gznode_->Subscribe(bat_topic, &AssemblyStatePublisher::UpdateBattery, this);
+  impl_->pump_sub_ = impl_->gznode_->Subscribe(pum_topic, &AssemblyStatePublisher::UpdatePump, this);
+  impl_->regulator_sub_ = impl_->gznode_->Subscribe(reg_topic, &AssemblyStatePublisher::UpdateRegulator, this);
+  impl_->sensor_sub_ = impl_->gznode_->Subscribe(sen_topic, &AssemblyStatePublisher::UpdateSensor, this);
+
+  const gazebo_ros::QoS & qos = impl_->ros_node_->get_qos();
+  rclcpp::QoS pub_qos = qos.get_publisher_qos("~/out", rclcpp::SensorDataQoS().reliable());
+  impl_->status_pub_ = impl_->ros_node_->create_publisher<ariac_msgs::msg::AssemblyState>("/ariac/" + model_name + "_assembly_state", pub_qos);
+
+  impl_->rate_ = 10;
+
+  impl_->battery_attached_ = false;
+  impl_->pump_attached_ = false;
+  impl_->regulator_attached_ = false;
+  impl_->sensor_attached_ = false;
+
+  impl_->timer_.Reset();
+  impl_->timer_.Start();
+
+  // Create a connection so the OnUpdate function is called at every simulation
+  // iteration. Remove this call, the connection and the callback if not needed.
+  impl_->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin(
+    std::bind(&AssemblyStatePublisher::OnUpdate, this));
+}
+
+void AssemblyStatePublisher::OnUpdate()
+{
+  auto elapsed = impl_->timer_.GetElapsed();
```

Comment:
This timer is neat, I didn't know it exists. Do you know if it works off of wall time or simulation time?

Reply:
This one's wall time